stl/inc/memory

```diff
@@ -1578,20 +1762,211 @@ private:
+        _Maybe_rebind_alloc_t<_Alloc, _STD remove_cv_t<_STD remove_all_extents_t<_Ty>>> _Alty(this->_Get_val());
+        _Uninitialized_value_construct_md_n_al(_Storage._Value, _STD extent_v<_Ty>, _Alty);
+    }
+
+    explicit _Ref_count_bounded_array_alloc(const _Alloc& _Al_arg, const _STD remove_extent_t<_Ty>& _Val)
+        : _Ebco_base<_Alloc>(_Al_arg), _Ref_count_base() {
+        _Maybe_rebind_alloc_t<_Alloc, _STD remove_cv_t<_STD remove_all_extents_t<_Ty>>> _Alty(this->_Get_val());
+        if constexpr (_STD rank_v<_Ty> > 1) {
+            for (size_t _It = 0; _It < _STD extent_v<_Ty>; ++_It)
+                _Uninitialized_copy_md_al(_Val, _Storage._Value[_It], _Alty);
+        } else {
+            _Uninitialized_fill_n(_Storage._Value, _STD extent_v<_Ty>, _Val, _Alty);
+        }
+    }
+
+    union {
+        _Wrap<_Ty> _Storage;
+    };
+
+private:
+    ~_Ref_count_bounded_array_alloc() {
+        // nothing to do; _Storage._Value already destroyed by _Destroy()
+    }
+
+    virtual void _Destroy() noexcept override { // destroy managed resource
+        _Maybe_rebind_alloc_t<_Alloc, _STD remove_all_extents_t<_Ty>> _Alty(this->_Get_val());
+        _Reverse_destroy_md_n_al(_Storage._Value, _STD extent_v<_Ty>, _Alty);
+    }
+
+    virtual void _Delete_this() noexcept override { // destroy self
+        _Rebind_alloc_t<_Alloc, _Ref_count_bounded_array_alloc> _Al(this->_Get_val());
+        this->~_Ref_count_bounded_array_alloc();
+        _Deallocate_plain(_Al, this);
+    }
+};
+#endif // _HAS_CXX20
+
 // FUNCTION TEMPLATE make_shared
 template <class _Ty, class... _Types>
-_NODISCARD shared_ptr<_Ty> make_shared(_Types&&... _Args) { // make a shared_ptr
+#if _HAS_CXX20 && __cpp_concepts
+requires(!_STD is_array_v<_Ty>)
+#endif // _HAS_CXX20
+    _NODISCARD shared_ptr<_Ty> make_shared(_Types&&... _Args) { // make a shared_ptr to non-array object
     const auto _Rx = new _Ref_count_obj2<_Ty>(_STD forward<_Types>(_Args)...);
     shared_ptr<_Ty> _Ret;
     _Ret._Set_ptr_rep_and_enable_shared(_STD addressof(_Rx->_Storage._Value), _Rx);
     return _Ret;
 }
 
+#if _HAS_CXX20 && __cpp_concepts
+template <class _Ty>
+requires _STD is_unbounded_array_v<_Ty> _NODISCARD shared_ptr<_Ty> make_shared(
+    size_t _Count) { // make a shared_ptr to an unbounded array
+    using _Refc = _Ref_count_unbounded_array<_Ty>;
+    using _Storage =
+        _STD aligned_storage_t<(_STD max)(alignof(_Ty), alignof(_Refc)), (_STD max)(alignof(_Ty), alignof(_Refc))>;
+    constexpr size_t _Rx_size_div_by_storage_size = (_STD max)(sizeof(_Refc) / sizeof(_Storage), size_t(1));
+    constexpr double _Ty_size_div_by_storage_size =
+        static_cast<double>(sizeof(_STD remove_extent_t<_Ty>)) / static_cast<double>(sizeof(_Storage));
+    const auto _Rx = reinterpret_cast<_Refc*>(
+        new _Storage[static_cast<size_t>(ceil(_Count * _Ty_size_div_by_storage_size)) + _Rx_size_div_by_storage_size]);

```

Comment:
This is using array `new` for `_Storage`, which could allocate extra memory to remember the number of `_Storage` objects. (In the Microsoft ABI, I believe that it doesn't because `_Storage` is plain old data, but I'm not 100% sure about this.) @CaseyCarter, IIRC you've worked with aligned allocations before; is there a better strategy?

Reply:
Part 1 of the new strategy appears to work; resolving now.